### ISSUE TITLE
Define usecases and workflows to support

### ISSUE THREAD

**Issue by bpoldrack** [psyinfra/onyo#335]
This issue is supposed to track discussion of what we think are usecases and workflows that are relevant for onyo.
From this, test cases defining them should then be extracted and implemented.

Primarily, I'd want to know from @aghisla, @j-leo and @aqw: 
- What are real-world scenarios that would make you want to interact with the inventory system?
- What information do you start with?
- What information needs to be discovered?
- What information needs to be put into it at what point in time?
- What would be typical/expected cases of some asset maybe already existing in the system, but additional information needs to be added to it or changed?

Some (partially) made-up example: My move to D&P.
I had a laptop, keyboard+mouse and a display and was assigned to a room. I took laptop, keyboard+mouse, moved to a different office and left the display in the old one.
Now, I imagine something like this: Maybe all those assets were associated with me before.
You'd now start with my name, name of group I left and name of group I joined.
So, you'd probably need to:
- move my association with groups first
- possibly after some communication move room association
- assets would need to split: keep display with former room and/or group, the rest stays (and moves) with me

^ This would need further ironing. When would such a split happen? Prior or after move of person? May be simultaneously? Is such a split expected to be a real-world usecase? (It wouldn't be necessary if the display was associated with the room or group to begin with. But then again: The same may be true for the keyboard. So, may be it's actually two moves: 1. Me (inlc. Laptop) and 2. keyboard (from group to group).

When we figure out these things, we should get to a couple of intended workflows, that can be tested and serve to set priorities for 1.0

**Comment by j-leo** [comment]
OK, so, your move:
- your old screen stayed in your old office ( no change to that, since screens are bound to rooms [exception: home office screens are bound to users]))
- your new screen was brought from the warehouse to your office, so its room number changed
- your laptop is still your laptop - no change to that
- keyboards, mice, network stands will keep to not be monitored

**Comment by aghisla** [comment]
Here is a list of common inventory-related actions that we want to track:

Additions to the inventory:

- add a new object to the inventory and the metadata available at the instant of delivery (metadata depending on the object, but the addition to inventory must happen at delivery, not later)
- add a new member of INM-7 (for any administrative value of "member", as long as they receive something that we inventorise)
- (to be discussed, until we get a working plan): migrate current inventory status from GLPI. Not a requirement that it will be automated (data in there are of very various levels of reliability and as we have to cherrypick data it becomes impossible to script)

Active inventory operations:

- reserve objects: they are in the ware-house but already booked for someone
- deploy objects: assign objects to members or rooms and hand out the actual object
- for objects that receive an inventory sticker from FZJ (which can happen weeks/months after delivery and usually after the object has been assigned and handed out): add the information about the sticker
- get objects back from members or rooms, ready to be redeployed
- get objects back for maintenance/repair (note: replacement parts may be tracked separately especially if seen as independent purchases)
- get objects back for retirement (includes extra paperwork)

Active queries on the inventory:

- where is a specific object (in the ware-house? assigned to someone? in some office onsite? at home? dead? out for repair?); the question comes usually from ourselves but can come from Admin (very rare, outside of the inventory reviews mentioned below). Admin will query either via an FZJ inventory number, sometimes a purchase order number; for us it's the hostname of a laptop, its FZJ inventory number, or the person currently owning it
- do we have a spare object of some type available? (most often, a laptop or screen in the ware-house s.l., filtering by metadata; sometimes parts for juseless)
- keep track of what is out for repair (repairs are rare and repairs with delays/issues are even rarer, so more of a double check)
- full list of inventorised objects currently assigned to a member (for user-leaving issues mostly)

Review/statistics:

- remote ware-house check (not much needed when onsite though)
- history of an object (where it has been before, assigned to whom)
- history of the assignments to a member (including moves across groups)
- Admin review: check every N years where every FZJ-stickered object of our Kostenstelle is located and if it still exists: having the means to quickly get an updated list is extremely useful
- Admin review also include some subset of software but that is not yet in the immediate scope of onyo, and what Admin wants to know is if that software is still active, nothing about who uses it or where

All for now! There is more hidden in each point but the overview should be good.

**Comment by aqw** [comment]
My perspective is mostly about auditing or forecasting:

- Starting with a purchase order, where is the FZJ-inventoried equipment resulting from it?
  - the purchasing system (ledger) should have an inventory number for each device, which allows us to follow it to the inventory system
  - e.g. `onyo get --filter serial=<inventory-number> full-path` (or whatever we choose to call those path-based pseudo keys)
- Forecast costs of user equipment for the next 2 years:
  - I cannot control vendor pricing nor the number of new hires. But I can anticipate the retirement of equipment.
  - I query all laptops (displays, headsets, etc) that are older than 3 years old that are not in the "retired" folder, and sort them by age.
    - e.g. `onyo get --filter type=laptop build-date < 20200101 type make model serial` (assuming that `build-date` is a key in the assets data; alternatively, the `commit-date` (future) pseudo-key could be abused)
    - I am comfortable with running one query for each `type` in this case, but  `||` is also necessary (and should be supported via python `re` IIRC).
- Determine consistency/integrity of repo data:
  - see validation
  - `onyo get --filter type=laptop build-date=<unset> full-path` (I want to see if any laptops are missing the `build-date` key)
- Apple issues a recall on the battery of all 13" MacBook Pros built between 2016-2017. We need to a list to coordinate which machines are affected, and where they are.
    - e.g. `onyo get --filter type=laptop build-date >= 20160101 build-date <= 20171231 full-path`

**Comment by bpoldrack** [comment]
@aghisla: Thanks much, that's pretty informative already!

> migrate current inventory status from GLPI. Not a requirement that it will be automated (data in there are of very various levels of reliability and as we have to cherrypick data it becomes impossible to script)

FTR: Seems to me it would still be worth scripting the import of the data as is, commit as starting point and do the cherry-picking in a separate commit. Gives a clear history that shows where data was coming from and how/why it was changed.

> reserve objects: they are in the ware-house but already booked for someone

Thinking out loud:
That seems an interesting case, because the solutions I see for this each come with a potential to complicate queries:
- have a `reserved` underneath `warehouse`, move the asset there and have the commit message reference the person/room/group/whatever-it-was-reserved-for
- have the association spelled out by subfolders underneath `reserved`.
Both approaches would make it a little more complicated to query the inventory for what is associated with a given name, if this is supposed to include the reserved assets. That's because it's not obvious to me, when to join things (and where to look for the condition - commit vs folder name).
More generally, that also points to an issue when there are several folders that are supposed to belong to the same entity (person or not). Hypothetical scenario: Someone gets a contract that assigns part time to group1 and part time to group2. That may not be likely in INM-7, but generally such scenarios do exist. I suppose a report would ideally still be able to merge the two. And then there's the related issue, that a name of a person isn't guaranteed to be unique. Only the full path would be, which would preclude such merges, unless explicitly instructed.
So, may be a merge should never be auto-discovered, but a query command needs to able to be told what to merge?
WDYT, @aqw?

**Comment by aqw** [comment]
> reserve objects: they are in the ware-house but already booked for someone

IMO, this is entirely up to how a site uses Onyo. We're taking a filesystem approach, so we don't do tags, we do categories. There are limitations of such an approach, but it's a choice, made at the very foundation of Onyo (filesystem). If we try to bolt tagging on, it will be an endless source of frustration, because it violates the filesystem approach.

Ultimately, it is a site's decision about what should be recorded in an inventory system vs not. If a site wants to "reserve" laptops that are in their warehouse, and they feel that this must be tracked in the inventory system rather than with post-its on a laptop, then they will need to create the appropriate folder structure under "warehouse" to do so. And that will make the history more noisy. But that is a desired outcome, because the decision was made that this is so important that it should be tracked in the inventory system. If the information is not important enough for us to want to retain history of it, then it should not be tracked IMO.

TL;DR: this is a user decision, not an Onyo design decision. The trade-offs need to be made by the user.

**Comment by bpoldrack** [comment]
@aqw 

Ok, I agree with that take. There may still be utility in some form of requesting a merge for a report, but I guess right now that's not something to focus on.

**Comment by aghisla** [comment]
As an extension that deserves a separate comment, rereading myself I notice how the core idea is to make Onyo move things around while I am the one knowing what it means. That's why I would use the Onyo concept of `onyo mv` for basically everything, and use the freetext (part of the) comment to give the real life context of the action.

**Comment by bpoldrack** [comment]
So, let me try to summarize and point out some things requiring clarification:

Asset operations:
-----------------

- Add a new asset
- Move an asset
- possibly remove an asset

What's not entirely clear to me, is deletion. I suppose in order to properly track things one would want to record a real-world disappearance of whatever nature as a move into "retired"/"lost"/"whatever".
However, removal would be a valid approach as well. It would have implications for how to query, though. That's because querying files that used to be there is (potentially) a lot more expensive over time.
So, removal probably largely for assets with some sort of data entry issue.

Requirements for those operations are the identifying metadata (pseudo-keys) that make up an asset's file name.
Strictly: For addition of an asset that information needs to be available at minimum.
For other operations, it could be obtained via a query that's based on different criteria.

Similarly, we have operations on locations/users:

- Add a location/user (folder)
- Move a location/user (folder), including its content
  Is this always including the content? Or should there be a safe-guard switch that says: "Only move if empty" or "Only move if no assets/files, but do include subfolders (same applies those, though)"?
- Probably: Remove a location/user
  Not clear to me, what this entails under what circumstances, though. What happens to folders and assets underneath?
  Strict requirement to be empty first?

Querying the inventory:
-----------------------

- Where is an asset?
  This may be based on different information to identify said asset:
  - FZJ inventory number
  - Purchase order number
    I suppose, this would be used to query ledger first, which in turn yields, well - what exactly, @aqw, @aghisla?
  - Hostname of a laptop
  - pseudo key(s)
  - @aghisla: You also mentioned to query for a specific asset and identify by "the person currently owning it".
    Not sure what you mean by that. Sounds more like "Give me all assets of a person, that also match the following criteria".
    So, basically the same kind of query as the above mentioned, but limiting the search to a person (folder), right?
- What assets match certain criteria?
  - pseudo keys
  - missing keys
    While I think most of this is indeed addressable by a dedicated query criterion (@aqw suggested something like `<unset>`),
    it seems to me, it would be nice to have a `--report-missing` switch with the query command, that would report on assets that miss any key that is used for matching in the given query.
    What do you think about this? For piping or whatever other sort of processing one would probably not want to pollute the output with this by default, but as an option could be nice.

Some desired matching and more generally the query command itself comes with some (potential) complexity. So, here are couple of questions:
- @aqw, you mentioned `onyo get --filter serial=<inventory-number> full-path`. So, `serial number == inventory number`? Was that intended? If so, I will need some idea when this is supposed to apply.
- Quite a bunch of suggested usecases rely on querying for the FZJ inventory number. Are there any criteria when exactly an FZJ inventory number needs to exist? Asking, because may be that would help figuring out some configurable validation rules.
- Related: What inventory number is ledger reporting when asked for a purchase order? Is there an INM-7 inventory number that would need to be distinguished?
- @aqw suggested a couple of more complex queries, like:
  - `onyo get --filter type=laptop build-date < 20200101 type make model serial`
    I suppose it would need to be `build-date<20200101` or `"build-date < 20200101"` in order to make it one argument.
    Furthermore, this looks like we need a `--` or similar that allows to separate between arguments to the `--filter` switch and the output keys.
    Alternatively, specify `--filter` multiple times, so: `onyo get --filter type=laptop --filter "build-date < 20200101" type make model serial`
    WDYT?
  - "but || is also necessary (and should be supported via python re IIRC": Not sure about syntax ATMm, but yes - should be doable
  - `onyo get --filter type=laptop build-date=<unset> full-path`
    Rather than reserving special values, what about `--missing build-date` instead?
  - `onyo get --filter type=laptop build-date >= 20160101 build-date <= 20171231 full-path`
    Do you prefer two criteria over a compact `"20160101 <= build-date <= 20171231"`? Not that important on the surface, but may depend on how and what from this query is build in reality.
    Generally, I think with those comparisons we see a limitation to CLI usage (not just this query).
    We don't know a type of value for a given YAML key (and I suppose we do not actually want to impose the need to provide full specifications onto users).
    Hence, we can't tell whether `20160101` is supposed to be string, integer or date. And therefore `<=` isn't clear.
    Currently, I think this would be something for a python API, where the user would actually provide "20160101" as an object of the respective type rather than always as a string via CLI.
    Are there more elaborate thoughts on that topic already?

History
-------

The requirements for reporting on history are not yet clear to me.
@aghisla brought up to inspect history of
- an asset
  What would need to be easily accessible here? I suppose, moves and commit messages I need to catch the eye immediately. What about the committer?
  What kind of question would the report need to provide the answer for? Does anyone imagine to feed the output into something else? What would that be?
- a user (folder)
  That's a tricky one. First off, I am not sure this is something an inventory system needs to be able to do, since it's about assets, not about managing users.
  However, there's certainly utility in it. Hence, if we can (easily) provide such a thing, we probably should.
  But: How? Main problem seems to be to match folder identities. How do we know a folder was moved? Git only tracks files.
  We'd need to reverse-engineer file moves to figure out whether that looks in fact like a folder move. Alternatively, I have already seen an issue about recording ony commands used in commit messages.
  This could be a (conceptually) easy way to do that. Report on usage of onyo commands rather than on git history. But that's still a little vague - when does what approach apply? Should those be different queries/command-modes?
  Would there be differences in how the output should look like?
  Any thoughts on that?

**Comment by aghisla** [comment]
Many thoughts!

So in order:

**Remove an asset**: as we (as ourselves and on request from Admin) do need to keep track of items for at least a few months after the physical object is gone, I choose to keep the asset file in a "retired" folder with potentially more subfolders by category (broken; lost; stolen...). I see more disadvantages in deleting the asset files and folders. That can however be done as a form of periodical archiving in a form that is easily searchable.

**Move a location/user** is in both cases moving a folder. Onyo doesn't need to know what the folder represents and I am comfortable, for our use cases, to let Onyo move the folder and all its contents (as it does for multifile assets represented by folders). The preview of the commit should be enough to make sure the action is semantically what is intended. I can't think of cases where we need to move an empty folder around, or check that the folder is empty before moving it. If there is nothing in the folder, the folder can go. I'm ok with a folder representing a room to be nuked when there is no equipment in there, it will be recreated next time we have equipment there. About removing a user folder, I'm ok as well for it to be gone when the user has no equipment anymore. One can think of doing some kind of archiving as for past equipment though.

**Query assets**: purchase number comes from various sources, the most complete is the purchasing repo's list of merge requests (for the metadata and documents) and ledger for the purchase date and cost. For your other question about "identifying by the person owning it" is indeed a query of assets assigned to someone and possibly filtered by type, though not that necessary as people rarely own more than a screenful of objects. As a note, queries are logically slightly different depending on who should receive the information (which tends to connect to the kind of information is used for the query).

**FZJ inventory numbers**: those are assigned to objects according to their price. There is no validation scheme that runs across the years and not even across the same model across years, because if it got over or below the price threshold it will or won't get the sticker. We also have assets with HHU inventory numbers whose rules are not known to me but I guess they are similar. I assume that all queries regarding FZJ inventory numbers will involve assets that surely have them and I would not use them for basically anything else. There is no extra INM-7 inventory number (God forbids!) and I don't plan to introduce yet one more scheme :D In ledger (see several purchases in ledger in 2021, each line is an object and it can have a FZJ inventory number as comment, but there is no connection with any specific object of that purchase (there is no other information in ledger for it. Not sure if we want to keep doing that).

**History**: for an asset it would be the (commit) history. Even plain output from git would be nice, at best interactive with tig, but at request. It's mostly us wanting to know where a machine has been last, or what was its status at some given time. I don't imagine to give it to any other tool yet, apart from grep, but that doesn't mean it should not be planned for some piping. Tracking a user may be indeed done by grepping for the commits containing the user's name (FSVO it). I wonder if it's simpler to consider folder/user history as a series of exports/archives rather than a product of git history as we can do for assets/files. Still need to think about this a bit.

Hope that clarifies some of the points!

ps. for our specific use cases I don't know how much more is OK to share here on this public repo, but definitely we need a place to track down information in a written and organised form.

**Comment by bpoldrack** [comment]
Thanks, @aghisla 

> As a note, queries are logically slightly different depending on who should receive the information (which tends to connect to the kind of information is used for the query).

Can you provide concrete examples for those differences?

> In ledger (see several purchases in ledger in 2021, each line is an object and it can have a FZJ inventory number as comment, but there is no connection with any specific object of that purchase (there is no other information in ledger for it. Not sure if we want to keep doing that).

Two things:
- again concretely: What do the lines of output look like? What I am aiming at here: How would one pipe one thing into the other. IOW: What would we want to be able to read from stdin?
- Overall this statement is a bit confusing to me. @aqw said, that he'd get the inventory numbers for a purchase order from ledger and now you seem to contradict this by "there is no connection with any specific object of that purchase". It's not clear to me what to make of that.

> for our specific use cases I don't know how much more is OK to share here on this public repo

As long as it doesn't contain real data, I don't see an issue. But we do need concrete test cases. Which is why I keep poking about details.

**Comment by aqw** [comment]
> Purchase order number
> I suppose, this would be used to query ledger first, which in turn yields, well - what exactly, @aqw, @aghisla?

This information is currently stored as comments in ledger. So the query would be some for of `vim` or `grep`.

The key is that it's discoverable with low effort.

> it seems to me, it would be nice to have a --report-missing switch with the query command, that would report on assets that miss any key that is used for matching in the given query.

Hmmm. I can see some use here, but I would reflect this back on you and ask to show some use cases. ;-P

> @aqw, you mentioned onyo get --filter serial=<inventory-number> full-path. So, serial number == inventory number? Was that intended? If so, I will need some idea when this is supposed to apply.

Yes, the idea is that "serial" is a unique identifier. Whenever there is an FZJ inventory number, we'd use that. When not, we'd use the vendor's serial. When none, we'll use a faux serial.

There are other solutions, of course. Nothing precludes Always using vendor serials and faux serials, and then inventory numbers in the file. Etc. Options are flexible, and Onyo is agnostic to that.

> I suppose it would need to be build-date<20200101 or "build-date < 20200101" in order to make it one argument.

Good point. Whatever it is, should match the syntax for `=`.

> Alternatively, specify --filter multiple times, so: onyo get --filter type=laptop --filter "build-date < 20200101" type make model serial
> WDYT?

I dislike the verbosity, but it may be needed. I think the appropriate answer here can only be made in conjunction with your following point, which is: what is the syntax for `||`.

I hope that helps. I skimmed through the text, and looked for places where I was `@`ed. If there's anything else you specifically want my input on, just ping me. :-)

**Comment by bpoldrack** [comment]
@aqw 
> Yes, the idea is that "serial" is a unique identifier. Whenever there is an FZJ inventory number, we'd use that. When not, we'd use the vendor's serial. When none, we'll use a faux serial.

Ok, I see how this is technically valid, of course, but given above usecase provided by @aghisla : "Report on all FZJ-stickered assets" - how would you do that, if there's no dedicated key for it? Or is "FZJ-stickered" and "has an FZJ inventory number" not the same? And even if so, is it not a usecase to filter all assets for having a FZJ inventory number? I don't think we can do that if we merge different notions of identifiers into one (pseudo) key.

**Comment by aqw** [comment]
> @aqw
> 
> > Yes, the idea is that "serial" is a unique identifier. Whenever there is an FZJ inventory number, we'd use that. When not, we'd use the vendor's serial. When none, we'll use a faux serial.
> 
> Ok, I see how this is technically valid, of course, but given above usecase provided by @aghisla : "Report on all FZJ-stickered assets" - how would you do that, if there's no dedicated key for it? Or is "FZJ-stickered" and "has an FZJ inventory number" not the same? And even if so, is it not a usecase to filter all assets for having a FZJ inventory number? I don't think we can do that if we merge different notions of identifiers into one (pseudo) key.
> 
> Edit: One could, of course, use a scheme that would prefix a FZJ inventory number somehow, so one can query for serials with that prefix, but that's sort of cheating. This approach just creates implicit keys inside values while pretending it's not. I think this is just making things harder to query by rejecting "structured records" of that info. Also How do you ensure consistency with that prefixing?

Fair point. And that is a valid use-case to generate a full list of what we think we have, vs what FZJ says we have.

I havn't thought about the design, but my gut instinct is to use regex (which I believe get supports) to define the serial pattern. Or use an FZJ prefix as you suggest.

**Comment by aqw** [comment]
IMO, this was a great issue to kickstart conversation and bring clarity to how Onyo was imagined to be used. Most of these ideas have been either implemented or have made it into issues.

The focus now is bringing those ideas to reality. Any future ones with have their own dedicated issue rather than continuing in this high-level discussion.
